waku/v2/peermanager/peer_manager.go

```diff
@@ -241,6 +243,98 @@ func (pm *PeerManager) Start(ctx context.Context) {
 		go pm.peerEventLoop(ctx)
 	}
 	go pm.connectivityLoop(ctx)
+	go pm.peerStoreLoop(ctx)
+}
+
+func (pm *PeerManager) peerStoreLoop(ctx context.Context) {
+	t := time.NewTicker(prunePeerStoreInterval)
+	defer t.Stop()
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-t.C:
+			pm.prunePeerStore()
+		}
+	}
+}
+
+func (pm *PeerManager) prunePeerStore() {
+	peers := pm.host.Peerstore().Peers()
+	numPeers := len(peers)
+	if numPeers < pm.maxPeers {
+		return
+	}
+	peerCntBeforePruning := numPeers
+	pm.logger.Debug("peerstore capacity exceeded, hence pruning", zap.Int("capacity", pm.maxPeers), zap.Int("numPeers", peerCntBeforePruning))
+
+	for _, peerID := range peers {
+		connFailues := pm.host.Peerstore().(wps.WakuPeerstore).ConnFailures(peerID)
+		if connFailues > maxFailedAttempts {
+			// safety check so that we don't end up disconnecting connected peers.
+			if pm.host.Network().Connectedness(peerID) == network.Connected {
+				pm.host.Peerstore().(wps.WakuPeerstore).ResetConnFailures(peerID)
+				continue
+			}
+			pm.host.Peerstore().RemovePeer(peerID)
+			numPeers--
+		}
+		if numPeers < pm.maxPeers {
+			pm.logger.Debug("finished pruning peer store", zap.Int("capacity", pm.maxPeers), zap.Int("beforeNumPeers", peerCntBeforePruning), zap.Int("afterNumPeers", numPeers))
+			return
+		}
+	}
+
+	notConnectedPeers := pm.getPeersBasedOnconnectionStatus("", network.NotConnected)
+	peersByTopic := make(map[string]peer.IDSlice)
+
+	if pm.RelayEnabled {
```

Comment:
Peers that are discovered via dnsdisc and peerExchange also populate the peerstore. Should we prune those as well? I just wonder if a light client wouldnt end up with an infinitely growing peerstore

Reply:
good point, i added this check in the end. 
because i don't think we include pubsubTopics while adding peers in lightclient. 
will confirm this once and if we are indeed specifying pubsubTopics, then this check can be removed.

discovered peers will get pruned if they are not connected to.
let me review this from lightclient perspective once again...will try to run this using status-desktop and see what is the behaviour

Reply:
updated in new commit https://github.com/waku-org/go-waku/pull/1193/commits/d044e56111a2795ad217bddf7f497481466b436c